foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/queries/DatabaseQueryMechanism.java

```diff
@@ -215,7 +219,14 @@ public boolean isJPQLCallQueryMechanism() {
      * This should be overridden by subclasses.
      * @return the row count.
      */
-    public abstract Integer executeNoSelect() throws DatabaseException;
+    public abstract Object executeNoSelect() throws DatabaseException;
```

Comment:
On a first look, this is sort of the change (Int -> Object) I don't like. I'm not saying it's wrong but isn't there some other alternative like ie introducing new method returning the right type instead of generic Object? From the diff, it looks like the return value will always be either Integer or DatabaseCall and having Object there may lead to headaches during maintenance..

Reply:
Sure, I could just create new methods, copy everything over, and even deprecate the old ones. However, I figured this was safe because it was internal API (also, this is HEAD where backwards compatibility isn't necessarily too important).

I went with the implementation that required the least amount of code duplication. It isn't that the overall execution strategy of INSERT queries is changing, just that it isn't necessarily ALWAYS returning an Integer up the call stack. Creating new API methods to call would be cleaner, but determining when to make those calls over the current defaults is trickier.

If it's a hard pass, I can take a look at creating new methods and what that would entail.

Reply:
I'm not sure yet, I'm still checking the codebase if there is something solving similar problem like handling multiple resultsets through one call and then somehow merging results already - what caught my attention is the comment about `DatasourceCall.NO_RETURN` - looking at [prepareInsertObject](https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/queries/DatasourceCallQueryMechanism.java#L612)  it seems to me, that at the point the query is being prepared, it may be already known that identity is used, so why not start by properly configuring the call itself?

Reply:
[#350487](https://bugs.eclipse.org/bugs/show_bug.cgi?id=350487) added handling of multiple resultsets through one call, so that might be good source

Reply:
> looking at [prepareInsertObject](https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/queries/DatasourceCallQueryMechanism.java#L612) it seems to me, that at the point the query is being prepared, it may be already known that identity is used

The order of operations is roughly this:
1) `NativeSequence` is created. It determines that IDENTITY is being used and that the Platform class supports GeneratedKeys. So, `NativeSequence.shouldUseGeneratedKeysIfPlatformSupports` is set `true`
2) InsertObjectQuery is created. It checks the descriptors sequence (possibly being the NativeSequence from step 1) and sets `DatabaseQuery.shouldReturnGeneratedKeys` during prepare.
3) A SQLCall is created for the InsertObjectQuery. That SQLCall is prepared, which copies the `DatabaseQuery.shouldReturnGeneratedKeys` to `DatabaseCall.shouldReturnGeneratedKeys`
4) The InsertObjectQuery is executed..
5) The SQLCall is cloned and executed.
6) The SQLCall clone creates a Statement and executes it.
7) The GeneratedKeys ResultSet is stored in the cloned SQLCall, which is returned up to `DatasourceCallQueryMechanism.insertObject()`
8) The SQLCall is passed down into `DatabaseQueryMechanism.updateObjectAndRowWithSequenceNumber()` where it reads the key from the ResultSet as the ID value
9) The ID value is passed to `ObjectBuilder.assignSequenceNumber()`, where the Sequence is not used, but rather the value passed in is

> at the point the query is being prepared, it may be already known that identity is used

Right, so during query prepare, we can know that. We could change the `org.eclipse.persistence.internal.databaseaccess.DatasourceCall.returnType` from `DatasourceCall.NO_RETURN` to a new type. I tried implementing that, but I didn't see much of a benefit and it may potentially involve more code changes. One issue is that at a couple places in the code, it appears to be assumed that `call.isNothingReturned()` == INSERT. That logic would need to be addressed.

I'm not saying this is the wrong fix or a bad idea, but I opted for the implementation that tried to reuse the INSERT execution code path, rather than creating a new parallel path.